zaps/0002-encrypt-push-notifications.md

```diff
@@ -0,0 +1,182 @@
+    push_key: SymmetricKey # Shared key between client and server.
+  ```
+- On the push bouncer, add a new table (parallel to `RemotePushDeviceToken`):
+  ```
+  class zilencer.models.RemotePushDevice:
+    device_id: RandomID[32]  # Unique key for this record, generated by bouncer.
+    realm: ForeignKey[RemoteRealm]
+    token_kind: "apns" | "fcm"
+    token: str  # Provided by client.
+    push_account_id: RandomID[32] # Generated by client; need not be unique.
+    expired_time: datetime | None
+  ```
+
+- We probably also need a `zerver.models.push_notifications.PushDevice`
+  for Zulip Cloud use with no third-party bouncer? Or we could make Cloud use a bouncer.
+
+- Additionally, the push bouncer has an asymmetric key `bouncer_key`,
+  of which the public part is baked into clients.
+
+## Operation
+
+### To register a new push device for a Zulip account
+
+1. The client checks the [`/api/v1/server_settings`](https://zulip.com/api/get-server-settings)
+   to see whether the server advertises the `encrypted_push_notifications`
+   capability (or has a new enough feature level, details TBD).
+
+   If it does not, it uses the legacy push notifications registration flow.
+   (Note: We’ll likely want settings to require E2E encryption; that can come
+   after the feature is complete).
+
+1. The client generates a new RandomID `push_account_id`. It builds a JSON
+   object `push_registration` of type `PushRegistration`, and encrypts it to
+   `bouncer_key` as `encrypted_push_registration`.
+
+1. The client makes an API call to the server (authenticated with the user’s
+   API key) with arguments `push_account_id` and `encrypted_push_registration`.
+
+   The server generates a symmetric key `push_key` and creates a `PushDevice`
+   instance. The server replies to the client with `push_key`, and asynchronously
+   initiates the next step.
+
+   - Next step asynchronous: It ensures that the bouncer service 500ing will just
+     cause queueing of those requests for the local server, the mobile client
+     doesn’t need to deal with retries here. Handling of 500 is discussed below.
+
+   - TBD: Optionally, the client could include some randomness to be mixed into
+     `push_key`. We can decide if this is needed when reviewing the exact crypto
+     primitives we end up using.
+
+1. The server makes an API call to the push bouncer (authenticated with the
+   server’s API key) with arguments `user_id`, `user_uuid`, `realm_uuid`,
+   `push_account_id`, and `encrypted_push_registration`.
+
+   The push bouncer:
+   1. Decrypts `push_registration`.
+
+   1. Checks the `timestamp` for liveness, probably 60s is a good age limit.
+      A coded `JsonableError` should be used if the timestamp is stale.
+
+   1. Checks that it doesn’t already have a `RemotePushDevice` for this `token`
+      and `push_account_id`. This check helps to avoid creating a duplicate
+      record if the client retries when a previous attempt actually succeeded.
```

Comment:
OTOH if it does already have such a record, it returns success with that `device_id`. In particular that’s the way we expect things to be if the remote server previously made the request, didn’t get the response (due to network failure or whatever), and retried.

Reply:
Should we make it:
```
 Checks that it doesn’t already have a `RemotePushDevice` for this `token` and `push_account_id`. If it does 
 already have such a record, it returns success with the `device_id` corresponding to that record. This helps to 
 avoid creating a duplicate record if the server previously made the request, didn’t get the success response 
 (due to network failure or whatever), and retried.
```

The "client retries" would be handled by the remote server to check if `bouncer_device_id` already exists ? -- If it doesn't, check if `push_account_id` exists ? ( It means, the async step to register with bouncer is still WIP). If none of them exists, then registration can take place.

Reply:
That update LGTM.

I don't quite follow the question after it, though. I think you're asking about what the server does when the client retries a registration request?

Reply:
I just meant to say --> "Server will handle things when the client retries a registration request, so no need to document it to sound like bouncer handles it (which we're doing currently)." The last para was about an overview of the various checks that the remote server can make when client-retries.

Reply:
Functionally no change, if that is deleted. 

It can just help with a better response message: "Already registered" vs "Registration in progress".